### PR TITLE
JSDocs

### PR DESCRIPTION
**Pull request by anurse**
* Changes `index.ts` to export symbols by name rather than using `export *`. This is the pattern used by libraries like [rxjs](https://github.com/ReactiveX/rxjs/blob/master/src/index.ts). It has the advantage of allowing you to export something to other modules in the package but still have the chance to hide them by not exporting them from `index`.
* JSDoc everywhere! IntelliSense in VS Code shows the docs.
* Tidied up some overloads to make for better IntelliSense

### CODE REVIEW

File: clients/ts/signalr/src/HttpClient.ts [hunk 1]
```diff
@@ -5,30 +5,89 @@ import { AbortSignal } from "./AbortController";
...
+
+    /** The time to wait for the request to complete before throwing a TimeoutError. Measured in milliseconds. */
     timeout?: number;
 }
 
+/** Represents an HTTP response. */
 export class HttpResponse {
+    /** Constructs a new instance of {@link HttpResponse} with the specified status code.
+     *
+     * @param {number} statusCode - The status code of the response.
```
Comment (JamesNK) [inline clients/ts/signalr/src/HttpClient.ts]: How come these param docs are separated by a dash?

File: clients/ts/signalr/src/HubConnectionBuilder.ts [hunk 2]
```diff
@@ -54,13 +89,21 @@ export class HubConnectionBuilder {
         return this;
     }
 
+    /** Configures the {@link HubConnection} to use the specified Hub Protocol
```
Comment (JamesNK) [inline clients/ts/signalr/src/HubConnectionBuilder.ts]: fullstop

File: clients/ts/signalr/src/HubConnectionBuilder.ts [hunk 3]
```diff
@@ -54,13 +89,21 @@ export class HubConnectionBuilder {
...
+     * @param {IHubProtocol} protocol The {@link IHubProtocol} implementation to use.
+     */
     public withHubProtocol(protocol: IHubProtocol): HubConnectionBuilder {
         Arg.isRequired(protocol, "protocol");
 
         this.protocol = protocol;
         return this;
     }
 
+    /** Creates a {@link HubConnection} from the configuration options specified in this builder
```
Comment (JamesNK) [inline clients/ts/signalr/src/HubConnectionBuilder.ts]: fullstop

File: clients/ts/signalr/src/IHttpConnectionOptions.ts [hunk 4]
```diff
@@ -5,11 +5,37 @@ import { HttpClient } from "./HttpClient";
...
+    /** Configures the logger used for logging
+     *
+     * Provide an {@link ILogger} instance, and log messages will be logged via that instance. Alternatively, provide a value from
+     * the {@link LogLevel} enumeration and a default logger which logs to the Console will be configured to log messages of the specified
+     * level (or higher).
+     */
     logger?: ILogger | LogLevel;
-    accessTokenFactory?: () => string | Promise<string>;
+
+    /** A function that provides an access token required for HTTP Bearer authentication
```
Comment (JamesNK) [inline clients/ts/signalr/src/IHttpConnectionOptions.ts]: fullstop

File: clients/ts/signalr/src/IHubProtocol.ts [hunk 5]
```diff
@@ -4,69 +4,162 @@
 import { ILogger } from "./ILogger";
 import { TransferFormat } from "./ITransport";
 
-export const enum MessageType {
+/** Defines the type of a Hub Message */
```
Comment (JamesNK) [inline clients/ts/signalr/src/IHubProtocol.ts]: fullstops all the docs

File: clients/ts/signalr/src/IHubProtocol.ts [hunk 6]
```diff
@@ -4,69 +4,162 @@
...
 }
 
+/** Defines properties common to all Hub messages relating to a specific invocation. */
 export interface HubInvocationMessage extends HubMessageBase {
+    /** A {@link MessageHeaders} dictionary containing headers attached to the message. */
     readonly headers?: MessageHeaders;
+    /** The ID of the invocation relating to this message.
+     *
+     * This is expected to be present for {@link StreamInvocationMessage} and {@link CompletionMessage}. It may
+     * be 'undefined' for a {@link InvocationMessage} if the sender does not expect a response.
```
Comment (JamesNK) [inline clients/ts/signalr/src/IHubProtocol.ts]: a -> an

File: clients/ts/signalr/src/Stream.ts [hunk 7]
```diff
@@ -7,17 +7,45 @@
 // depend on RxJS in the core library, so instead we duplicate the minimum logic needed and then users can easily adapt these into
 // proper RxJS observables if they want.
 
+/** Defines the expected type for a {@link IStreamSubscriber}.
```
Comment (JamesNK) [inline clients/ts/signalr/src/Stream.ts]: a -> an

File: clients/ts/signalr/src/HttpClient.ts (hunk 1 above)
Comment (anurse) [reply clients/ts/signalr/src/HttpClient.ts]: It's "optional" but I thought it was required initially. I'll tidy that up.

File: clients/ts/signalr/src/HubConnection.ts [hunk 8]
```diff
@@ -22,6 +23,11 @@ export class HubConnection {
     private timeoutHandle: NodeJS.Timer;
     private receivedHandshakeResponse: boolean;
 
+    /** The server timeout in milliseconds.
+     *
+     * If this timeout elapses without receiving any messages from the server, the connection will be terminated with an error.
+     * The default timeout value is 30 seconds (30,000 milliseconds).
```
Comment (JamesNK) [inline clients/ts/signalr/src/HubConnection.ts]: nit: reverse these. The default value is 30,000

File: clients/ts/signalr/src/HubConnection.ts [hunk 9]
```diff
@@ -183,7 +229,19 @@ export class HubConnection {
...
+    /** Removes all handlers for the specified hub method.
+     *
+     * @param {string} methodName The name of the method to remove handlers for.
+     */
+    public off(methodName: string): void;
+
+    /** Removes the specified handler for the specified hub method.
+     *
+     * @param {string} methodName The name of the method to remove handlers for.
+     * @param {Function} method The handler to remove. This must be the same Function instance as the one passed to {@link on}.
```
Comment (JamesNK) [inline clients/ts/signalr/src/HubConnection.ts]: Maybe add a note that passing in a function that wasn't passed to on will have no effect.

**Comment by anurse** [comment]
:up: :date: Fixed build issues and retargeted to `release/2.1`. We want to try and get this and the XML docs in before the other PRs so that we can easily see what we've done so far by diffing without seeing the docs PRs as noise.

File: clients/ts/signalr/src/IHubProtocol.ts [hunk 10]
```diff
@@ -4,69 +4,162 @@
...
+    /** Indicates the message is a Ping message and implements the {@link PingMessage} interface */
     Ping = 6,
+    /** Indicates the message is a Close message and implements the {@link CloseMessage} interface */
     Close = 7,
 }
 
-export interface MessageHeaders { [key: string]: string; }
+/** Defines a dictionary of string keys and string values representing headers attached to a Hub message. */
+export interface MessageHeaders {
+    /** Gets or sets the header with the specified key */
```
Comment (JamesNK) [inline clients/ts/signalr/src/IHubProtocol.ts]: fullstop

File: clients/ts/signalr/src/IHubProtocol.ts (hunk 5 above)
Comment (JamesNK) [reply clients/ts/signalr/src/IHubProtocol.ts]: still missing a fullstop

File: clients/ts/signalr/src/IHubProtocol.ts [hunk 11]
```diff
@@ -4,69 +4,162 @@
...
+    [key: string]: string;
+}
 
-export type HubMessage = InvocationMessage | StreamInvocationMessage | StreamItemMessage | CompletionMessage | CancelInvocationMessage | PingMessage | CloseMessage;
+/** Union type of all known Hub messages */
```
Comment (JamesNK) [inline clients/ts/signalr/src/IHubProtocol.ts]: fullstop

File: clients/ts/signalr/src/JsonHubProtocol.ts [hunk 12]
```diff
@@ -7,16 +7,30 @@ import { TransferFormat } from "./ITransport";
 import { NullLogger } from "./Loggers";
 import { TextMessageFormat } from "./TextMessageFormat";
 
-export const JSON_HUB_PROTOCOL_NAME: string = "json";
+const JSON_HUB_PROTOCOL_NAME: string = "json";
```
Comment (JamesNK) [inline clients/ts/signalr/src/JsonHubProtocol.ts]: This makes it private to the file?

File: clients/ts/signalr/src/index.ts [hunk 13]
```diff
@@ -2,14 +2,15 @@
...
+export { DefaultHttpClient, HttpClient, HttpRequest, HttpResponse } from "./HttpClient";
+export { IHttpConnectionOptions } from "./IHttpConnectionOptions";
+export { HubConnection } from "./HubConnection";
+export { HubConnectionBuilder } from "./HubConnectionBuilder";
+export { MessageType, MessageHeaders, HubMessage, HubMessageBase, HubInvocationMessage, InvocationMessage, StreamInvocationMessage, StreamItemMessage, CompletionMessage, PingMessage, CloseMessage, CancelInvocationMessage, IHubProtocol } from "./IHubProtocol";
+export { ILogger, LogLevel } from "./ILogger";
+export { HttpTransportType, TransferFormat, ITransport } from "./ITransport";
+export { IStreamSubscriber, IStreamResult, ISubscription } from "./Stream";
+export { NullLogger } from "./Loggers";
+export { JsonHubProtocol } from "./JsonHubProtocol";
```
Comment (JamesNK) [inline clients/ts/signalr/src/index.ts]: I like being explicit.
Need to be careful that nothing has disappeared in the change.

File: clients/ts/signalr/src/JsonHubProtocol.ts (hunk 12 above)
Comment (anurse) [reply clients/ts/signalr/src/JsonHubProtocol.ts]: Correct.

File: clients/ts/signalr/src/index.ts (hunk 13 above)
Comment (anurse) [reply clients/ts/signalr/src/index.ts]: I did a pretty careful scan to make sure everything that was public before is still public.

File: clients/ts/signalr/src/HttpClient.ts [hunk 14]
```diff
@@ -5,30 +5,86 @@ import { AbortSignal } from "./AbortController";
...
     constructor(statusCode: number, statusText: string, content: string);
+
+    /** Constructs a new instance of {@link HttpResponse} with the specified status code, message and binary content.
+     *
+     * @param {number} statusCode The status code of the response.
+     * @param {string} statusText The status message of the response.
+     * @param {ArrayBuffer} content The content of the response.
+     */
     constructor(statusCode: number, statusText: string, content: ArrayBuffer);
+
```
[the remark this answers is not in the document]
Comment (anurse) [reply clients/ts/signalr/src/HttpClient.ts]: Not needed but I'll remove the new line. When you have overloads, the actual implementation function signature isn't used for anything, not docs, not IntelliSense, etc.

File: clients/ts/signalr/src/HubConnection.ts [hunk 15]
```diff
@@ -165,6 +206,11 @@ export class HubConnection {
         return p;
     }
 
+    /** Registers a handler that will be invoked when the hub method with the specified method name is invoked.
+     *
+     * @param {string} methodName The name of the hub method to define.
+     * @param {Function} newMethod The handler that will be raised when the hub method is invoked.
```
[the remark this answers is not in the document]
Comment (anurse) [reply clients/ts/signalr/src/HubConnection.ts]: Yes, we call it a "hub method" in the C# client too.

File: clients/ts/signalr/src/IHttpConnectionOptions.ts [hunk 16]
```diff
@@ -5,11 +5,37 @@ import { HttpClient } from "./HttpClient";
...
+
+    /** A boolean indicating if message content should be logged.
+     *
+     * Message content can contain sensitive user data, so this is disabled by default.
+     */
     logMessageContent?: boolean;
+
+    /** A boolean indicating if negotiation should be skipped.
+     *
+     * Negotiation can only be skipped when the {@link transport} property is set to 'HttpTransportType.WebSockets'
```
Comment (BrennanConroy) [inline clients/ts/signalr/src/IHttpConnectionOptions.ts]: Fullstop!!.

File: clients/ts/signalr/src/IHubProtocol.ts [hunk 17]
```diff
@@ -4,69 +4,162 @@
...
 
+/** A hub message sent to request that a streaming invocation be canceled. */
 export interface CancelInvocationMessage extends HubInvocationMessage {
+    /** @inheritDoc */
     readonly type: MessageType.CancelInvocation;
+    /** The invocation ID. */
+    readonly invocationId: string;
 }
 
+/** A protocol abstraction for communicating with SignalR hubs.  */
```
Comment (BrennanConroy) [inline clients/ts/signalr/src/IHubProtocol.ts]: **H**ubs?

**Pull request merged**